File: clsMSSpectrum.vb
Option Strict On

''' <summary>
''' Used to track the m/z and intensity values of a given mass spectrum
''' </summary>
Public Class clsMSSpectrum

    ''' <summary>
    ''' Scan number
    ''' </summary>
    ''' <returns></returns>
    ''' <remarks>0 if not in use</remarks>
    Public Property ScanNumber As Integer

    Public ReadOnly Property IonCount As Integer
        Get
            Return IonsMZ.Count
        End Get
    End Property

    ''' <summary>
    ''' List of m/z's
    ''' </summary>
    Public ReadOnly IonsMZ As List(Of Double)

    ''' <summary>
    ''' List of intensities
    ''' </summary>
    Public ReadOnly IonsIntensity As List(Of Double)

    ''' <summary>
    ''' Constructor that only sets the scan number
    ''' </summary>
    Public Sub New(intScanNumber As Integer)
        ScanNumber = intScanNumber

        IonsMZ = New List(Of Double)
        IonsIntensity = New List(Of Double)
    End Sub

    ''' <summary>
    ''' Constructor that sets the scan number and stores m/z and intensity data (float intensities)
    ''' </summary>
    Public Sub New(intScanNumber As Integer, mzList As IList(Of Double), intensityList As IList(Of Single), dataCount As Integer)
        Me.New(intScanNumber)

        For i = 0 To dataCount - 1
            IonsMZ.Add(mzList(i))
            IonsIntensity.Add(intensityList(i))
        Next
    End Sub

    ''' <summary>
    ''' Constructor that sets the scan number and stores m/z and intensity data (double intensities)
    ''' </summary>
    Public Sub New(intScanNumber As Integer, mzList As IList(Of Double), intensityList As IList(Of Double), dataCount As Integer)
        Me.New(intScanNumber)

        For i = 0 To dataCount - 1
            IonsMZ.Add(mzList(i))
            IonsIntensity.Add(intensityList(i))
        Next
    End Sub

    ''' <summary>
    ''' Clear the m/z and intensity values (but leave the scan number unchanged)
    ''' </summary>
    Public Sub Clear()
        IonsMZ.Clear()
        IonsIntensity.Clear()
    End Sub

    ''' <summary>
    ''' Clear the m/z and intensity values, and update the scan number
    ''' </summary>
    Public Sub Clear(newScanNumber As Integer)
        IonsMZ.Clear()
        IonsIntensity.Clear()
        ScanNumber = newScanNumber
    End Sub

    Public Function Clone() As clsMSSpectrum
        Return Copy(Me)
    End Function

    Public Function Copy(objSource As clsMSSpectrum) As clsMSSpectrum
        Dim newSpectrum = New clsMSSpectrum(objSource.ScanNumber, objSource.IonsMZ, objSource.IonsIntensity, objSource.IonsMZ.Count)
        Return newSpectrum
    End Function

    Public Sub ReplaceData(spectrum As clsMSSpectrum, scanNumberOverride As Integer)

        ScanNumber = spectrum.ScanNumber
        If ScanNumber <> scanNumberOverride Then
            ScanNumber = scanNumberOverride
        End If

        IonsMZ.Clear()
        IonsMZ.AddRange(spectrum.IonsMZ)

        IonsIntensity.Clear()
        IonsIntensity.AddRange(spectrum.IonsIntensity)
    End Sub

    Public Sub ShrinkArrays(ionCountNew As Integer)
        If ionCountNew > IonsMZ.Count Then
            Throw New Exception("ShrinkArrays should only be called with a length less than or equal to the current length")
        End If

        Dim countToRemove = IonsMZ.Count - ionCountNew
        If countToRemove = 0 Then Exit Sub

        IonsMZ.RemoveRange(ionCountNew, countToRemove)
        IonsIntensity.RemoveRange(ionCountNew, countToRemove)
    End Sub

End Class

File: clsMSSpectrum.cs
using System;
using System.Collections.Generic;

namespace MASIC
{
    /// <summary>
    /// Used to track the m/z and intensity values of a given mass spectrum
    /// </summary>
    public class clsMSSpectrum
    {
        /// <summary>
        /// Scan number
        /// </summary>
        /// <remarks>0 if not in use</remarks>
        public int ScanNumber { get; set; }

        /// <summary>
        /// Number of items in IonsMZ or IonsIntensity
        /// </summary>
        public int IonCount => IonsMZ.Count;

        /// <summary>
        /// List of m/z's
        /// </summary>
        public readonly List<double> IonsMZ;

        /// <summary>
        /// List of intensities; should be the same length as IonsMZ
        /// </summary>
        public readonly List<double> IonsIntensity;

        /// <summary>
        /// Constructor that only sets the scan number
        /// </summary>
        public clsMSSpectrum(int scanNumber)
        {
            ScanNumber = scanNumber;

            IonsMZ = new List<double>();
            IonsIntensity = new List<double>();
        }

        /// <summary>
        /// Constructor that sets the scan number and stores m/z and intensity data (float intensities)
        /// </summary>
        public clsMSSpectrum(int scanNumber, IList<double> mzList, IList<float> intensityList, int dataCount)
            : this(scanNumber)
        {
            IonsMZ.Capacity = dataCount;
            IonsIntensity.Capacity = dataCount;
            for (var i = 0; i < dataCount; i++)
            {
                IonsMZ.Add(mzList[i]);
                IonsIntensity.Add(intensityList[i]);
            }
        }

        /// <summary>
        /// Constructor that sets the scan number and stores m/z and intensity data (double intensities)
        /// </summary>
        public clsMSSpectrum(int scanNumber, IList<double> mzList, IList<double> intensityList, int dataCount)
            : this(scanNumber)
        {
            IonsMZ.Capacity = dataCount;
            IonsIntensity.Capacity = dataCount;
            for (var i = 0; i < dataCount; i++)
            {
                IonsMZ.Add(mzList[i]);
                IonsIntensity.Add(intensityList[i]);
            }
        }

        /// <summary>
        /// Clear the m/z and intensity values (but leave the scan number unchanged)
        /// </summary>
        public void Clear()
        {
            IonsMZ.Clear();
            IonsIntensity.Clear();
        }

        /// <summary>
        /// Clear the m/z and intensity values, and update the scan number
        /// </summary>
        public void Clear(int newScanNumber, int newCapacity = 0)
        {
            IonsMZ.Clear();
            IonsIntensity.Clear();
            if (newCapacity > 0)
            {
                var currentCapacity = IonsMZ.Capacity;
                if (newCapacity < currentCapacity / 2)
                {
                    currentCapacity = newCapacity;
                }
                IonsMZ.Capacity = Math.Max(currentCapacity, newCapacity);
                IonsIntensity.Capacity = Math.Max(currentCapacity, newCapacity);
            }
            ScanNumber = newScanNumber;
        }

        /// <summary>
        /// Make a deep copy of this mass spectrum
        /// </summary>
        // ReSharper disable once UnusedMember.Global
        public clsMSSpectrum Clone()
        {
            return Copy(this);
        }

        /// <summary>
        /// Make a deep copy of the given mass spectrum
        /// </summary>
        /// <param name="sourceSpectrum"></param>
        public clsMSSpectrum Copy(clsMSSpectrum sourceSpectrum)
        {
            var newSpectrum = new clsMSSpectrum(sourceSpectrum.ScanNumber, sourceSpectrum.IonsMZ, sourceSpectrum.IonsIntensity, sourceSpectrum.IonsMZ.Count);
            return newSpectrum;
        }

        /// <summary>
        /// Replace data stored in IonsMZ, IonsIntensity, and Scan Number
        /// </summary>
        /// <param name="spectrum"></param>
        /// <param name="scanNumberOverride"></param>
        public void ReplaceData(clsMSSpectrum spectrum, int scanNumberOverride)
        {
            ScanNumber = spectrum.ScanNumber;
            if (ScanNumber != scanNumberOverride)
            {
                ScanNumber = scanNumberOverride;
            }

            IonsMZ.Clear();
            IonsIntensity.Clear();

            if (IonsMZ.Capacity / 2 > spectrum.IonsMZ.Count)
            {
                IonsMZ.Capacity = spectrum.IonsMZ.Count;
                IonsIntensity.Capacity = spectrum.IonsIntensity.Count;
            }

            IonsMZ.AddRange(spectrum.IonsMZ);
            IonsIntensity.AddRange(spectrum.IonsIntensity);
        }

        /// <summary>
        /// Remove data from the end of IonsMZ and IonsIntensity
        /// </summary>
        /// <param name="ionCountNew"></param>
        public void ShrinkArrays(int ionCountNew)
        {
            if (ionCountNew > IonsMZ.Count)
            {
                throw new Exception("ShrinkArrays should only be called with a length less than or equal to the current length");
            }

            var countToRemove = IonsMZ.Count - ionCountNew;
            if (countToRemove == 0)
                return;

            IonsMZ.RemoveRange(ionCountNew, countToRemove);
            IonsIntensity.RemoveRange(ionCountNew, countToRemove);
        }
    }
}
